step03/application.py

```diff
@@ -0,0 +1,126 @@
+import logging
+import json
+import os
+import boto3
+from botocore.exceptions import ClientError
+import flask
+from flask import request, Response
+from io import BytesIO
+from pathlib import PurePath
+from urllib.parse import unquote_plus
+from PIL import Image
+
+BUCKET_NAME = 'gg-photo-bucket'
+OUTPUT_FOLDER = 'thumbnails'
+
+# Create and configure the Flask application
+application = flask.Flask(__name__)
+application.logger.setLevel(logging.INFO)
+
+session = boto3.Session(region_name='us-east-1')
+s3 = session.client('s3')
+
+
+class S3Message:
+  def __init__(self, key):
+    self.key = key
+
+  @classmethod
+  def parse(cls, message):
+    '''
+      This class method parses the request message from sqsd.
+      unquote_plus used to replace "+" with spaces from files uploaded
+
+      :returns: S3Message object
+    '''
+    # assumes only on record
+    record = message['Records'].pop()
+    key = unquote_plus(record['s3']['object']['key'])
+    return cls(key)
+
+def create_thumbnail(input_stream, size=(128, 128)):
+  '''
+    This function creates a thumbnail of an image.
+    It expects bytes and returns bytes.
+
+    :param input_stream:
+    :param size:
+    :returns: thumbnail output stream
+  '''
+
+  output_stream = BytesIO()
+  image = Image.open(BytesIO(input_stream))
+  image.thumbnail(size)
+  image.save(output_stream, image.format)
+  output_stream.seek(0)
+  return output_stream
+
+def create_thumbnail_key(key):
+  '''
+    This function creates the S3 key for the thumbnail to be uploaded to S3
+    based on the key of the original image.
+
+    :param key:
+    :returns: thumbnail key (path) in specified S3 bucket
+  '''
+
+  key_path = PurePath(key)
+  file_stem = key_path.stem
+  file_ext = key_path.suffix
+  return '{}/{}.thumbnail{}'.format(OUTPUT_FOLDER, file_stem, file_ext)
+
+
+@application.route('/', methods=['POST'])
+def process_message():
+  if not request.json:
+    return Response("No Message", status=415)
+
+  # Prints SQS Message body
+  application.logger.info(request.json)
+
+  # incoming request message from sqsd
+  message = S3Message.parse(request.json)
+
+  try:
+    get_object_response = s3.get_object(
+      Bucket=BUCKET_NAME,
+      Key=message.key
+    )
+  except ClientError as boto_error:
+    '''boto3 errors (s3)'''
+    print(f'Botocore Error: {boto_error}')
+
+  try:
+    stream = get_object_response['Body'].read()
+    thumbnail_stream = create_thumbnail(stream)
+    thumbnail_key = create_thumbnail_key(message.key)
+  except IOError as e:
```

Comment:
Probably should reraise the exception here since we want the API call to fail. Reevaluate each of the exception below whether or not they should cause failure

Reply:
I re-raised the exceptions for IOError, OSError, and KeyError since all of these would have similar issues and would not allow proper processing of the original image and creation of the thumbnail.